torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2262,7 +2365,83 @@ def _load_sharded_state_dict(
+                    args, kwargs = _to_kwargs(args, kwargs, self.compute_device.index, False)
+                    args = args[0]
+                    kwargs = kwargs[0]
+                    if self._mixed_precision_enabled_for_params():
+                        input_dtype = self.mixed_precision.param_dtype
+                        args, kwargs = self._cast_fp_inputs_to_precision(
+                            input_dtype, *args, **kwargs,
+                        )
+                handles = self._module_to_handles[module]
+                params_to_reshard = []
+                # TODO (awgu): Replace the below logic to check if the handles'
+                # parameters are not needed not just for this module but any
+                # subsequent module (right now, it may reshard too early)
+                for handle in self._handles:
+                    # TODO (awgu): Make into a set?
+                    if handle not in handles and handle.flat_param in self._unsharded_params:
+                        params_to_reshard.append(handle.flat_param)
+                # TODO (awgu): track which handles are unsharded...
+                reshard(params_to_reshard)
+                flat_params = [handle.flat_param for handle in handles]
+                for param in flat_params:
+                    assert param.device != torch.device("cpu"), \
+                        f"param {[n for n in param._prefixed_param_names]} on CPU!"
+                self._rebuild_full_params(flat_params, source="fwd")
+                torch.cuda.current_stream().wait_stream(self._streams["all_gather"])
+                self._register_post_backward_hooks(flat_params)
+                for handle in handles:
+                    handle._unflatten(as_params=False)
+                outputs = module_forward(*args, **kwargs)
+                # TODO (awgu): Normally, we would reshard here. However, maybe
+                # we only reshard in the next module's forward. If there is no
+                # next module, then that means this module is the last module
+                # in the execution order, in which case its parameters do not
+                # need to be resharded since they most likely will be
+                # immediately used for gradient computation. This mirrors the
+                # intention behind setting `reshard_after_forward=False` for
+                # the FSDP root.
+                outputs = self._register_pre_backward_hooks(outputs, flat_params)
+                self.training_state = TrainingState_.IDLE
+            return outputs
+
+        return patched_forward
+
+    def _unpatch_forwards(self):
+        """Unpatches ``module.forward()`` to remove additional FSDP logic."""
+        for module, forward in self._module_to_forward.items():
+            module.forward = forward
+
     def forward(self, *args: Any, **kwargs: Any) -> Any:
+        if self._use_exec_order_policy:
+            assert self._is_root is None or self._is_root
```

Comment:
and rather than or?

Reply:
If `_lazy_init()` has not been called yet (i.e. first iteration), then `self._is_root` is `None`.